`<utility>`: `pair::swap(const pair&)` interacts badly with `__declspec(dllexport)`

This is a regression caused by merging #2687 in VS 2022 17.3 Preview 2:

```
C:\Temp>type meow.cpp
```
```cpp
#include <utility>
using namespace std;

struct __declspec(dllexport) MyPair : pair<int, int> {};
```
```
C:\Temp>cl /EHsc /nologo /W4 /MD /c /std:c++20 meow.cpp
meow.cpp

C:\Temp>cl /EHsc /nologo /W4 /MD /c /std:c++latest meow.cpp
meow.cpp
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\utility(107): error C2672: 'swap': no matching overloaded function found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\type_traits(1965): note: could be 'void std::swap(_Ty (&)[_Size],_Ty (&)[_Size]) noexcept(<expr>)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\utility(107): note: 'void std::swap(_Ty (&)[_Size],_Ty (&)[_Size]) noexcept(<expr>)': could not deduce template argument for '_Ty (&)[_Size]' from '_Ty'
        with
        [
            _Ty=int
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\type_traits(1965): note: see declaration of 'std::swap'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\type_traits(1962): note: or       'void std::swap(_Ty &,_Ty &) noexcept(<expr>)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\utility(106): note: 'void std::swap(_Ty &,_Ty &) noexcept(<expr>)': could not deduce template argument for '_Enabled'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\type_traits(1962): note: see declaration of 'std::swap'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\utility(408): note: see reference to function template instantiation 'void std::_Swap_adl<const _Ty1>(_Ty &,_Ty &) noexcept(false)' being compiled
        with
        [
            _Ty1=int,
            _Ty=int
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\include\utility(406): note: while compiling class template member function 'void std::pair<int,int>::swap(const std::pair<int,int> &) noexcept(false) const'
meow.cpp(4): note: see reference to class template instantiation 'std::pair<int,int>' being compiled
```

Relevant code:

https://github.com/microsoft/STL/blob/2263d931f48a4e196036b2523af7a0048bb96b46/stl/inc/utility#L331-L339

WG21-N4910 [[pairs.pair]/45.2](https://eel.is/c++draft/pairs.pair#45.2) says:

> *Mandates:* For the second overload, `is_swappable_v<const T1>` is `true` and `is_swappable_v<const T2>` is `true`.

So we aren't supposed to constrain this, but that interacts badly with `dllexport` which tries to instantiate everything.

I suspect that we need the `template <int = 0>` workaround, also applied to `tuple`.

Originally reported as DevCom-10109884 and internal VSO-1586226 / AB#1586226 .

--- a/stl/inc/tuple
+++ b/stl/inc/tuple
@@ -582,6 +582,7 @@ public:
     }
 
 #if _HAS_CXX23
+    template <int = 0> // see GH-3013
     constexpr void swap(const tuple& _Right) const
         noexcept(conjunction_v<is_nothrow_swappable<const _This>, is_nothrow_swappable<const _Rest>...>) {
         _Swap_adl(_Myfirst._Val, _Right._Myfirst._Val);

--- a/stl/inc/utility
+++ b/stl/inc/utility
@@ -329,6 +329,7 @@ struct pair { // store a pair of values
     }
 
 #if _HAS_CXX23
+    template <int = 0> // see GH-3013
     constexpr void swap(const pair& _Right) const
         noexcept(is_nothrow_swappable_v<const _Ty1>&& is_nothrow_swappable_v<const _Ty2>) {
         if (this != _STD addressof(_Right)) {

--- a/tests/std/tests/GH_001638_dllexport_derived_classes/test.compile.pass.cpp
+++ b/tests/std/tests/GH_001638_dllexport_derived_classes/test.compile.pass.cpp
@@ -9,8 +9,10 @@
 #include <queue>
 #include <set>
 #include <stack>
+#include <tuple>
 #include <unordered_map>
 #include <unordered_set>
+#include <utility>
 #include <vector>
 
 #if _HAS_CXX20
@@ -50,3 +52,7 @@ struct __declspec(dllexport) ExportedStack : stack<int> {};
 struct __declspec(dllexport) ExportedSpan : span<int> {};
 struct __declspec(dllexport) ExportedSpanThree : span<int, 3> {};
 #endif // _HAS_CXX20
+
+// Test GH-3013 "<utility>: pair::swap(const pair&) interacts badly with __declspec(dllexport)"
+struct __declspec(dllexport) ExportedPair : pair<int, int> {};
+struct __declspec(dllexport) ExportedTuple : tuple<int, int, int> {};

